### PR TITLE
Replace binary/reasons configurations with a list of coding configurations

### PR DESCRIPTION
**Pull request by as2388**
This will give us the flexibility we need to support yes/no/reasons questions in the follow-ups later in the project. It also allows for greater flexibility in coding beyond supporting just yes/no/reasons should we need it in future, moves more of the configuration from random project files into pipeline_configuration.py, and simplifies some of the code in the pipeline.

When reviewing it will probably be easiest to look at pipeline_configuration.py first, as this is what the rest of the changes are implementing.

Adding @IsaackMwenda as a secondary reviewer, mainly as a sanity check on the proposed new pipeline_configuration.py.

### CODE REVIEW

File: src/analysis_file.py [hunk 1]
```diff
@@ -26,8 +27,13 @@ def td_has_stop_code(td, coding_plans):
...
         """
         for plan in coding_plans:
-            if plan.code_scheme.get_code_with_id(td[plan.coded_field]["CodeID"]).control_code == Codes.STOP:
-                return True
+            for cc in plan.coding_configurations:
+                if cc.coding_mode == CodingModes.SINGLE:
+                    if cc.code_scheme.get_code_with_id(td[cc.coded_field]["CodeID"]).control_code == Codes.STOP:
+                        return True
+                else:
+                    if td[f"{cc.analysis_file_key}{Codes.STOP}"] == Codes.MATRIX_1:
```
Comment (lukechurch) [inline src/analysis_file.py:35]: This looks a bit surprising as a compound key - could you give me an example of what it looks like?

Comment (lukechurch) [reply src/analysis_file.py:35]: Oh, are these for the export to the dataset file, that makes more sense.... If true, ignore above.

File: src/analysis_file.py [hunk 2]
```diff
@@ -152,31 +114,62 @@ def generate(user, data, csv_by_message_output_path, csv_by_individual_output_pa
...
+        # Convert codes to their string/matrix values
+        for td in data:
+            analysis_dict = dict()
+            for plan in PipelineConfiguration.RQA_CODING_PLANS + PipelineConfiguration.SURVEY_CODING_PLANS:
+                for cc in plan.coding_configurations:
+                    if cc.coding_mode == CodingModes.SINGLE:
+                        analysis_dict[cc.analysis_file_key] = \
+                            cc.code_scheme.get_code_with_id(td[cc.coded_field]["CodeID"]).string_value
+                    else:
+                        show_matrix_keys = []
```
Comment (lukechurch) [inline src/analysis_file.py:156]: assert MATRIX_MODE? (Sorry for the repeat need to do this, a language with safe enumerations would avoid it)

File: src/analysis_file.py [hunk 3]
```diff
@@ -152,31 +114,62 @@ def generate(user, data, csv_by_message_output_path, csv_by_individual_output_pa
...
+                        for code in cc.code_scheme.codes:
+                            show_matrix_keys.append(f"{cc.analysis_file_key}{code.string_value}")
+
+                        for label in td.get(cc.coded_field, []):
+                            analysis_dict[f"{cc.analysis_file_key}{cc.code_scheme.get_code_with_id(label['CodeID']).string_value}"] = Codes.MATRIX_1
```
Comment (lukechurch) [inline src/analysis_file.py]: This is getting too complex to be sensible as an inline key generation - please assign to an intermediary variable

File: src/analysis_file.py [hunk 4]
```diff
@@ -193,19 +186,21 @@ def generate(user, data, csv_by_message_output_path, csv_by_individual_output_pa
...
-
-                contains_non_nc_key = False
-                for key in matrix_keys:
-                    if key.startswith(plan.analysis_file_key) and not key.endswith(Codes.NOT_CODED) \
-                            and td.get(key) == Codes.MATRIX_1:
-                        contains_non_nc_key = True
-                if not contains_non_nc_key:
-                    td.append_data({f"{plan.analysis_file_key}{Codes.NOT_CODED}": Codes.MATRIX_1},
-                                   Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+            for plan in PipelineConfiguration.RQA_CODING_PLANS + PipelineConfiguration.SURVEY_CODING_PLANS:
```
Comment (lukechurch) [inline src/analysis_file.py:191]: Please discuss with me [again] if there are ways we could make these implicit decision algorithms read more straightforwardly. At the moment logic and data structure access code gets a bit blended together.

File: src/apply_manual_codes.py [hunk 5]
```diff
@@ -3,217 +3,74 @@
...
+                try:
+                    if path.exists(coda_input_path):
+                        f = open(coda_input_path, "r")
+
+                    if cc.coding_mode == CodingModes.SINGLE:
+                        TracedDataCodaV2IO.import_coda_2_to_traced_data_iterable(
+                            user, data, plan.id_field, {cc.coded_field: cc.code_scheme}, f)
+                    else:
+                        TracedDataCodaV2IO.import_coda_2_to_traced_data_iterable_multi_coded(
+                            user, data, plan.id_field, {cc.coded_field: cc.code_scheme}, f)
```
Comment (lukechurch) [inline src/apply_manual_codes.py:31]: Consider catching failures and logging a warning

**Comment by lukechurch** [comment]
@as2388 This code is inevitably hard to review. Please could you also outline what testing you done on it?

File: src/lib/code_imputation_functions.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import time
```
Comment (lukechurch) [inline src/lib/code_imputation_functions.py:1]: Good to separate these, thanks

**Review by lukechurch — APPROVED** [review]
I have no reason to hold this up, but it is challenging to review. Once we know what the testing has looked like, let's merge and then try again to think through how to manage the blend of logic and data structure access

File: src/analysis_file.py (hunk 2 above)
Comment (as2388) [reply src/analysis_file.py:156]: added

File: src/analysis_file.py (hunk 3 above)
Comment (as2388) [reply src/analysis_file.py]: done

**Comment by as2388** [comment]
> @as2388 This code is inevitably hard to review. Please could you also outline what testing you done on it?

I've run over the production and test data/codes for AHADI and confirmed that the outputs are equivalent. Most of the code changed here is not "new", it's just been moved/adapted to handle a list of code configurations.

The only things that haven't been possible are tests over fully labelled real-world datasets or on yes/no/reasons codes, because there wasn't time to test on e.g. ADSS first under the timelines that I thought I was under when this was written.

**Review by IsaackMwenda — APPROVED** [review]
LGTM! Great that you have separated code imputations too.

File: src/apply_manual_codes.py (hunk 5 above)
Comment (as2388) [reply src/apply_manual_codes.py:31]: Hmm, I don't think I want to log a warning here, I want it to die.

**Comment by as2388** [comment]
We now have coded real-world data for some RQAs and demogs on this project. Output analysis files are equivalent, so merging now.

**Pull request merged**